daemon/network.go

```diff
@@ -682,64 +682,74 @@ func buildPeerInfoResources(peers []networkdb.PeerInfo) []network.PeerInfo {
 	return peerInfo
 }
 
-func buildIpamResources(r *types.NetworkResource, nwInfo libnetwork.NetworkInfo) {
-	id, opts, ipv4conf, ipv6conf := nwInfo.IpamConfig()
+// buildIPAMResources constructs a [network.IPAM] from the network's
+// IPAM information for inclusion in API responses.
+func buildIPAMResources(nw libnetwork.NetworkInfo) network.IPAM {
+	var ipamConfig []network.IPAMConfig
 
-	ipv4Info, ipv6Info := nwInfo.IpamInfo()
+	ipamDriver, ipamOptions, ipv4Conf, ipv6Conf := nw.IpamConfig()
 
-	r.IPAM.Driver = id
-
-	r.IPAM.Options = opts
-
-	r.IPAM.Config = []network.IPAMConfig{}
-	for _, ip4 := range ipv4conf {
-		if ip4.PreferredPool == "" {
+	hasIPv4Config := false
+	for _, cfg := range ipv4Conf {
+		if cfg.PreferredPool == "" {
+			continue
+		}
+		hasIPv4Config = true
+		ipamConfig = append(ipamConfig, network.IPAMConfig{
+			Subnet:     cfg.PreferredPool,
+			IPRange:    cfg.SubPool,
+			Gateway:    cfg.Gateway,
+			AuxAddress: cfg.AuxAddresses,
+		})
+	}
+
+	hasIPv6Config := false
+	for _, cfg := range ipv6Conf {
+		if cfg.PreferredPool == "" {
 			continue
 		}
-		iData := network.IPAMConfig{}
-		iData.Subnet = ip4.PreferredPool
-		iData.IPRange = ip4.SubPool
-		iData.Gateway = ip4.Gateway
-		iData.AuxAddress = ip4.AuxAddresses
-		r.IPAM.Config = append(r.IPAM.Config, iData)
-	}
-
-	if len(r.IPAM.Config) == 0 {
-		for _, ip4Info := range ipv4Info {
-			iData := network.IPAMConfig{}
-			iData.Subnet = ip4Info.IPAMData.Pool.String()
-			if ip4Info.IPAMData.Gateway != nil {
-				iData.Gateway = ip4Info.IPAMData.Gateway.IP.String()
+		hasIPv6Config = true
+		ipamConfig = append(ipamConfig, network.IPAMConfig{
+			Subnet:     cfg.PreferredPool,
+			IPRange:    cfg.SubPool,
+			Gateway:    cfg.Gateway,
+			AuxAddress: cfg.AuxAddresses,
+		})
+	}
+
+	if !hasIPv4Config || !hasIPv6Config {
```

Comment:
Ah, right, so I had a quick look, and I think I kept these conditions to prevent calling `nw.IpamInfo()` if we don't need it. That function also locks the network, does a `CopyTo` to create (deep) copies of the structs. Still need to profile some of that, but I thought it would be a good thing to not call it if we don't need it.

> Also, it'd be great to document what `nw.IpamConfig()` and `nw.IpamInfo()` return and why we fallback to these values here.

Yes, we should. I could use some help on some of that (some of the parts are not yet clear to me either at this point, so need to untangle some of it). I'll leave that for a follow-up 😅 


https://github.com/moby/moby/blob/536e3692c631b4a8c8c2b82ddd17dc000f7fd3af/libnetwork/network.go#L1822-L1846

Reply:
That said, we should probably make `IpamInfo()` just return "one" IP class (IPv4 or IPv6) instead of always returning both. I'll look at that in a follow-up as well.